valhalla/common/telescope_states.py

```diff
@@ -15,116 +15,188 @@
+        if time_diff >= dt and event_source['reason'] in lump_data['reasons']:
+            return True
+
+        # If the event is an ENCLOSURE_INTERLOCK, and the lump includes a SEQUENCER_UNAVAILABLE, it is part of the
+        # lump. This is because the two states are redundant.
+        if event_source['type'] == 'ENCLOSURE_INTERLOCK' and 'SEQUENCER_UNAVAILABLE' in lump_data['types']:
+            return True
+
+        return False
+
+    def _categorize(self, lump):
+        # TODO: Categorize each lump in a useful way for network users.
+        event_type, event_reason = lump['types'][0], lump['reasons'][0]
+        for this_type, this_reason in zip(lump['types'], lump['reasons']):
+            # If the state in ENCLOSURE INTERLOCK, wait to instead categorize as SEQUENCER_UNAVAILABLE.
+            if this_type.upper() != 'ENCLOSURE_INTERLOCK':
+                event_type, event_reason = this_type, this_reason
+                break
+        return event_type, event_reason
+
+    def _lump_end(self, lump, next_event=None):
+        if not next_event or str(lump['telescope']) != str(self._telescope(next_event)):
+            return self.end
+        return self._es_to_datetime(next_event['timestamp'])
+
+    def _set_lump(self, event):
+        return {
+            'reasons': [event['_source']['reason']],
+            'types': [event['_source']['type']],
+            'start': self._es_to_datetime(event['_source']['timestamp']),
+            'telescope': self._telescope(event['_source']),
+            'latest_timestamp': self._es_to_datetime(event['_source']['timestamp'])
+        }
+
+    def _update_lump(self, lump, event):
+        if event['_source']['type'] not in lump['types'] or event['_source']['reason'] not in lump['reasons']:
+            lump['reasons'].append(event['_source']['reason'])
+            lump['types'].append(event['_source']['type'])
+        lump['latest_timestamp'] = self._es_to_datetime(event['_source']['timestamp'])
+        return lump
+
+    def get(self):
+        telescope_states = {}
+        current_lump = dict(reasons=None, types=None, start=None)
+
+        for event in self.event_data:
+            if self._telescope(event['_source']) not in self.available_telescopes:
+                continue
+
+            if current_lump['start'] is None:
+                current_lump = self._set_lump(event)
+                continue
+
+            if self._belongs_in_lump(event['_source'], current_lump):
+                current_lump = self._update_lump(current_lump, event)
+            else:
+                lump_end = self._lump_end(current_lump, event['_source'])
+                if lump_end < self.start:
+                    current_lump = self._set_lump(event)
+                    continue
+
+                telescope_states = self._update_states(telescope_states, current_lump, lump_end)
+                current_lump = self._set_lump(event)
+
+        if current_lump['start']:
```

Comment:
It will probably always be true. If it isn't true, that means that we didn't get any data from ES for some reason, in which case I would rather return an empty dictionary than crash.